packages/data-table-server/src/dataTableService/userDefined/UserDefinedDataTableService.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2022 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup } from '@tupaia/utils';
```

Comment:
See comment above about what the difference is? From a quick scan it looks like UserDefinedDataTables have specific shaped params, but nothing here makes me think this shouldn't just be on DataTable.

Reply:
So the general idea is that with `UserDefinedDataTables`, the user gets to control which parameter the data table takes. This is that part of the UI where the user creates the parameters. 

So in the backend, we basically have to read the config to work out what parameters the user has defined, and then create a yup schema from that config so that when we request data for that data table we can validate it has been requested with the correct parameters.

Whereas with the non-user defined data-tables, the parameters are hard coded. I did actually consider moving those hard-coded parameters into the config, and then the code could be the same with the user-defined data-tables? The main drawback I found there is that we couldn't use quite as precise validators, but perhaps it's worth it, thoughts?

Reply:
I see, I think I see it like configuring each DataTable is just a form with a few fields, similar to a Slack bot integration, or a VSCode plugin settings screen, or a chrome extension settings screen etc. It just so happens to be that the SqlDataTable has a field which is an array.

Like GoogleSheetsDataTable will need sheetId. 
```
{
  name: 'myGoogleSheetsDataTable',
  params: {
    sheetId: '123'
  }
}
```
so the UI for this would be 
```
Name: [       ]
Params:
  Sheet Id: [      ]
```

And SqlDataTable will need connectionId and sqlParameters (optional).
```
{
  name: 'mySqlDataTable',
  params: {
    connectionId: '1',
    sqlParams: {
      dataElements: ['ABC123', 'XYZ456']
    }
  }
}
```
and the UI for this would be:
```
Name: [       ]
Params:
  Connection: [      v]
  SQL Params (optional):
    1 Key [      ] Value [      ] Type [     v] (+)(-) 
    1 Key [      ] Value [      ] Type [     v] (+)(-) 
    1 Key [      ] Value [      ] Type [     v] (+)(-) 
```

Not sure if I'm off the mark but yeah just how the mental image is in my head.